lib/utils/compose_ts.ts

```diff
@@ -264,63 +318,161 @@ async function performResolution(
+
+	if (!(await fs.exists(fullDockerfilePath))) {
+		throw new ExpectedError(stripIndent`
+			Error: specified Dockerfile not found:
+			Specified dockerfile: "${fullDockerfilePath}"
+			Project's source folder: "${nativeProjectPath}"
+			Note that the specified Dockerfile path should be relative to the source folder.
 		`);
 	}
 
-	if (!existsSync(nativeDockerfilePath)) {
-		exitWithExpectedError(stripIndent`
-			Error: Dockerfile not found: "${nativeDockerfilePath}"
+	if (!contains(nativeProjectPath, fullDockerfilePath)) {
+		throw new ExpectedError(stripIndent`
+			Error: the specified Dockerfile must be in a subfolder of the source folder:
+			Specified dockerfile: "${fullDockerfilePath}"
+			Project's source folder: "${nativeProjectPath}"
 		`);
 	}
 
-	const { dir, ext, name } = parse(nativeDockerfilePath);
+	const { dir, ext, name } = path.parse(fullDockerfilePath);
 	if (ext) {
-		const nativePathMinusExt = join(dir, name);
-
-		if (existsSync(nativePathMinusExt)) {
-			exitWithExpectedError(stripIndent`
-				Error: "${name}" exists on the same folder as "${dockerfilePath}".
-				When an alternative Dockerfile name is specified, a file with the same
-				base name (minus the file extension) must not exist in the same folder.
-				This is because the base name file will be auto generated and added to
-				the tar stream that is sent to the docker daemon, resulting in duplicate
-				Dockerfiles and undefined behavior.
+		const nativePathMinusExt = path.join(dir, name);
+		if (await fs.exists(nativePathMinusExt)) {
+			throw new ExpectedError(stripIndent`
+				Error: "${name}" exists on the same folder as "${nativeDockerfilePath}".
+				When an alternative Dockerfile name is specified, a file with the same base name
+				(minus the file extension) must not exist in the same folder. This is because
+				the base name file will be auto generated and added to the tar stream that is
+				sent to balenaEngine or the Docker daemon, resulting in duplicate Dockerfiles
+				and undefined behavior.
 			`);
 		}
 	}
-	return posix.normalize(toPosixPath(dockerfilePath));
+	return toPosixPath(nativeDockerfilePath);
+}
+
+export interface ProjectValidationResult {
+	dockerfilePath: string;
+	registrySecrets: RegistrySecrets;
+}
+
+/**
+ * Perform "sanity checks" on the project directory, e.g. for the existence
+ * of a 'Dockerfile[.*]' or 'docker-compose.yml' file or 'package.json' file.
+ * Also validate registry secrets if any, and perform checks around an
+ * alternative specified dockerfile (--dockerfile) if any.
+ *
+ * Return the parsed registry secrets if any, and the "tar stream path" for
+ * an alternative specified Dockerfile if any (see validateSpecifiedDockerfile()).
+ */
+export async function validateProjectDirectory(
+	sdk: BalenaSDK,
+	opts: {
+		dockerfilePath?: string;
+		noParentCheck: boolean;
+		projectPath: string;
+		registrySecretsPath?: string;
+	},
+): Promise<ProjectValidationResult> {
```

Comment:
The `validateProjectDirectory()` function is the main feature of this PR, and reason for the 'minor' semver.